controlpanel/oidc.py

```diff
@@ -0,0 +1,44 @@
+from urllib.parse import urlencode
+
+from django.conf import settings
+from django.urls import reverse
+from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+
+from controlpanel.api.models import User
+
+
+class OIDCSubAuthenticationBackend(OIDCAuthenticationBackend):
+    """
+    Authentication backend which matches users by their `sub` claim
+    """
+
+    def create_user(self, claims):
+        user = User.objects.create(
+            pk=claims.get("sub"),
+            username=claims.get(settings.OIDC_FIELD_USERNAME),
+            email=claims.get(settings.OIDC_FIELD_EMAIL),
+            name=claims.get(settings.OIDC_FIELD_NAME),
+        )
+        user.save()
+        # user.aws_create_role()
+        # user.helm_create()
+        return user
+
+    def filter_users_by_claims(self, claims):
+        sub = claims.get("sub")
+        if not sub:
+            return self.UserModel.objects.none()
+
+        try:
+            return User.objects.filter(pk=sub)
+
+        except User.DoesNotExist:
+            return self.UserModel.objects.none()
```

Comment:
#### 3a. Recognise users by their Auth0 `sub` claim
>Lines 128 - 128

We're using the user's Auth0 `sub` claim (eg: `github|1234`) as their user ID, so I subclassed the OIDC authentication backend to handle this.

[Next →]()